crates/llm-remote-http/src/lib.rs

```diff
@@ -0,0 +1,186 @@
+use anyhow::Result;
+use reqwest::{
+    header::{HeaderMap, HeaderValue},
+    Client, Url,
+};
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use spin_core::async_trait;
+use spin_llm::{access_denied_error, LlmEngine};
+use spin_world::llm::{self as wasi_llm, InferencingParams};
+use std::collections::HashSet;
+
+#[derive(Clone)]
+pub struct RemoteHttpLlmEngine {
+    auth_token: String,
+    url: Url,
+    client: Option<Client>,
+    allowed_models: HashSet<String>,
+}
+
+#[derive(Serialize)]
+#[serde(rename_all(serialize = "camelCase"))]
+struct InferRequestBodyParams {
+    max_tokens: u32,
+    repeat_penalty: f32,
+    repeat_penalty_last_n_token_count: u32,
+    temperature: f32,
+    top_k: u32,
+    top_p: f32,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all(deserialize = "camelCase"))]
+struct InferUsage {
+    prompt_token_count: u32,
+    generated_token_count: u32,
+}
+#[derive(Deserialize)]
+struct InferResponseBody {
+    text: String,
+    usage: InferUsage,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all(deserialize = "camelCase"))]
+struct EmbeddingUsage {
+    prompt_token_count: u32,
+}
+
+#[derive(Deserialize)]
+struct EmbeddingResponseBody {
+    embeddings: Vec<Vec<f32>>,
+    usage: EmbeddingUsage,
+}
+
+#[async_trait]
+impl LlmEngine for RemoteHttpLlmEngine {
+    fn set_allowed_models(&mut self, models: HashSet<String>) {
+        self.allowed_models = models;
+    }
+
+    async fn infer(
+        &mut self,
+        model: wasi_llm::InferencingModel,
+        prompt: String,
+        params: InferencingParams,
+    ) -> Result<wasi_llm::InferencingResult, wasi_llm::Error> {
+        if !self.allowed_models.contains(&model) {
+            return Err(access_denied_error(&model));
+        }
```

Comment:
Can we move this check outside of the specific implementations and into `LlmDispatch`? It would be nice if we didn't have to repeat this check across all implementors.

Reply:
done!